### PR TITLE
explore/13-cost-model-vs-nn

### LINKED ISSUE #15: add costs
add cost values in the plan representation

### LINKED ISSUE #17: Cost model vs NN
Compare learned NN with cost model on plan ranking problem (in generalisation mode!): "What is the probability that random 2 plans will be ordered correctly via cost (NN prediction) comparison?"

### PR DESCRIPTION
**Pull request by zinchse**
**tldr;** NN is really good at generalising - its predictions are better than the cost model in both plans ranking and best plans selection.

# Plans ranking

## Real data
To verify how well NN predictions can answer the question "Is plan X faster than plan Y?" we measured Kendall and Spearman coefficients between a) execution time and predicted cost (cost model), b) execution time and predicted execution time (NN).

We obtained the previous results - in almost all cases, NN predictions correlate better than cost model predictions, except for `ood` data.

| data_type | Kendall [cost] | Kendall [NN] | Spearman [cost] | Spearman [NN] |
|-----------|----------------|--------------|-----------------|---------------|
| train/val | 0.435±0.000    | 0.634±0.010  | 0.585±0.000     | 0.765±0.010   |
| test      | 0.468±0.000    | 0.588±0.005  | 0.615±0.000     | 0.724±0.006   |
| ood       | 0.394±0.000    | 0.178±0.029  | 0.529±0.000     | 0.255±0.039   |

## Random data

Even experiments on random data confirmed both the importance of statistics and the presence of generalization through training its own cost model:

| data_type | noise_type | Kendall [cost] | Kendall [NN] | Spearman [cost] | Spearman [NN] |
|-----------|------------|----------------|--------------|-----------------|---------------|
| test      | ops        | 0.061±0.000    | 0.573±0.020  | 0.076±0.000     | 0.757±0.021   |
| test      | stats      | 0.061±0.000    | 0.412±0.031  | 0.076±0.000     | 0.576±0.040   |
| test      | all        | 0.061±0.000    | 0.219±0.016  | 0.076±0.000     | 0.322±0.020   |

# Plans selection

image
image
image

A separate study was also conducted to determine how useful the obtained predictions are for reducing the number of parameters that need to be explored. To do this, we selected the top-k hintsets according to the predicted cost and the NN predictions and chose the best among them. The results were also expected - on `train/val` and `test` datasets, NN predictions turned out to be significantly more useful, but on `ood` data, the cost model performed slightly better.

### CODE REVIEW

**Comment by zinchse** [comment]
ok; final step: figure out how many resources we need for a good generalisation in online scenario

**Comment by zinchse** [comment]
note that we compared the predictions of the NN with the costs for the plans that were obtained under the influence of the hints mechanism; it is necessary to keep in mind that the hints are applied by hardcoded overestimation of the cost
